src/optimizer/rule/pushdown_limit.rs

```diff
@@ -0,0 +1,310 @@
+                children: PatternChildrenPredicate::None,
+            }]),
+        }
+    };
+    static ref PUSH_LIMIT_INTO_TABLE_SCAN_RULE: Pattern = {
+        Pattern {
+            predicate: |op| match op {
+                Operator::Limit(_) => true,
+                _ => false
+            },
+            children: PatternChildrenPredicate::Predicate(vec![Pattern {
+                predicate: |op| match op {
+                    Operator::Scan(_) => true,
+                    _ => false
+                },
+                children: PatternChildrenPredicate::None,
+            }]),
+        }
+    };
+}
+
+/// Push down `Limit` past a `Project`.
+pub struct LimitProjectTranspose;
+
+impl Rule for LimitProjectTranspose {
+    fn pattern(&self) -> &Pattern {
+        &LIMIT_PROJECT_TRANSPOSE_RULE
+    }
+
+    fn apply(&self, node_id: HepNodeId, graph: &mut HepGraph) {
+        graph.swap_node(
+            node_id,
+            graph.children_at(node_id)[0]
+        );
+    }
+}
+
+/// Combines two adjacent Limit operators into one, merging the expressions into one single
+/// expression.
+pub struct EliminateLimits;
+
+impl Rule for EliminateLimits {
+    fn pattern(&self) -> &Pattern {
+        &ELIMINATE_LIMITS_RULE
+    }
+
+    fn apply(&self, node_id: HepNodeId, graph: &mut HepGraph) {
+        if let Operator::Limit(op) = graph.operator(node_id) {
+            let child_id = graph.children_at(node_id)[0];
+            if let Operator::Limit(child_op) = graph.operator(child_id) {
+                let new_limit_op = LimitOperator {
+                    offset: op.offset + child_op.offset,
+                    limit: cmp::min(op.limit, child_op.limit),
+                };
+
+                graph.remove_node(child_id, false);
+                graph.replace_node(
+                    node_id,
+                    OptExprNode::OperatorRef(
+                        Operator::Limit(new_limit_op)
+                    )
+                );
+            }
+        }
+    }
+}
+
+/// Add extra limits below JOIN:
+/// 1. For LEFT OUTER and RIGHT OUTER JOIN, we push limits to the left and right sides,
+/// respectively.
+/// 2. For INNER and CROSS JOIN, we push limits to both the left and right sides if join condition
+/// is empty.
+pub struct PushLimitThroughJoin;
+
+impl Rule for PushLimitThroughJoin {
+    fn pattern(&self) -> &Pattern {
+        &PUSH_LIMIT_THROUGH_JOIN_RULE
+    }
+
+    fn apply(&self, node_id: HepNodeId, graph: &mut HepGraph) {
+        let child_id = graph.children_at(node_id)[0];
+        let join_type = if let Operator::Join(op) = graph.operator(child_id) {
+            Some(op.join_type)
+        } else {
+            None
+        };
+
+        if let Some(ty) = join_type {
+            if let Some(grandson_id) = match ty {
+                JoinType::Left => Some(graph.children_at(child_id)[0]),
+                JoinType::Right => Some(graph.children_at(child_id)[1]),
+                _ => None
+            } {
```

Comment:
Miss inner/cross join

Reply:
inner/corss/full cannot be pushed down

Reply:
> Miss inner/cross/full join

inner/corss/full cannot be pushed down

Reply:
But why comment `For INNER and CROSS JOIN, we push limits to both the left and right sides if join condition is empty.`